app/src/main/java/com/celzero/bravedns/data/AppMode.kt

```diff
@@ -43,57 +49,77 @@ class AppMode internal constructor(private val context: Context,
                                    private val dnsCryptRelayEndpointRepository: DNSCryptRelayEndpointRepository,
                                    private val proxyEndpointRepository: ProxyEndpointRepository,
                                    private val persistentState: PersistentState) {
-    private val INVALID_DNS_TYPE = -1
-    private var appDNSMode: Long = PREF_DNS_INVALID
-    private val proxyTypeInternal = "Internal"
-    private val proxyTypeExternal = "External"
-    private var dnsType: Int = INVALID_DNS_TYPE
-    private var socks5ProxyEndpoint: ProxyEndpoint? = null
+    private var appDnsMode: Long = PREF_DNS_INVALID
 
     companion object {
         private var connectedDNS: MutableLiveData<String> = MutableLiveData()
         var cryptRelayToRemove: String = ""
+
+        fun isFirewallActive(mode: TunnelMode): Boolean {
+            return mode.dnsMode == Settings.DNSModeNone && mode.firewallMode != Settings.BlockModeNone
+        }
+    }
+
+    init {
+        connectedDNS.postValue(persistentState.connectedDnsName)
     }
 
+    // TODO: Move the app state to BraveVpnService.State- (NEW, WORKING, FAILING, PAUSED).
+    enum class AppState(val state: Int) {
+        PAUSED(10), ACTIVE(11)
+    }
+
+    data class TunnelMode(val dnsMode: Long, val firewallMode: Long, val proxyMode: Long)
 
-    fun getDNSMode(): Long {
-        if (appDNSMode != PREF_DNS_INVALID) return appDNSMode
+    fun getFirewallMode(): Long {
+        return persistentState.firewallMode.toLong()
+    }
+
+    fun setFirewallMode() {
+        persistentState.firewallMode = determineFirewallMode().toInt()
+    }
+
+    fun getDnsType(): Int {
+        return persistentState.dnsType
+    }
+
+    suspend fun getDnsMode(): Long {
+        if (appDnsMode == PREF_DNS_INVALID) {
+            setDnsMode()
+        }
 
-        when (persistentState.dnsType) {
+        return appDnsMode
+    }
+
+    fun setDnsMode() {
+        appDnsMode = determineDnsMode()
+    }
+
+    private fun determineDnsMode(): Long {
+        // Case: app mode - firewall, DNS mode should be none.
+        if (persistentState.braveMode == APP_MODE_FIREWALL) {
+            return Settings.DNSModeNone
+        }
+
+        // app mode - DNS & DNS+Firewall mode
+        return when (persistentState.dnsType) {
             PREF_DNS_MODE_DOH -> {
-                appDNSMode = if (persistentState.allowDNSTraffic) {
+                if (persistentState.allowDNSTraffic) {
                     Settings.DNSModePort
                 } else {
                     Settings.DNSModeIP
                 }
             }
             PREF_DNS_MODE_DNSCRYPT -> {
-                appDNSMode = Settings.DNSModeCryptPort
+                Settings.DNSModeCryptPort
```

Comment:
see above. creating a new enum and using `when` clause would make sure all cases are handled. Here, `allowDNSTraffic` is being ignored. Same for dns-proxy below.

Reply:
done. `allowDNSTraffic` is considered approp, now.